From 64f09110bda24b92ec8db4a2cfe03784b86a9d41 Mon Sep 17 00:00:00 2001
From: Fugang Duan <b38611@freescale.com>
Date: Wed, 22 Jul 2015 12:42:40 +0800
Subject: [PATCH 0016/5242] MLK-11274 net: fec: add mii bus up_failed flag to
 reflect the real status

commit  ab841a45cf286da866f3ea60457d6a5c9ad88789 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add mii bus up_failed flag to reflect the real mii bus status.

Signed-off-by: Fugang Duan <B38611@freescale.com>
Reported-and-tested-by: Zhang Sanshan <B51434@freescale.com>
(cherry picked from commit: ea348e597501d44841a28d8ee099361e89d63520)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/freescale/fec.h      |    1 +
 drivers/net/ethernet/freescale/fec_main.c |    6 +++++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/fec.h b/drivers/net/ethernet/freescale/fec.h
index 0b2da4c..4c5c691 100644
--- a/drivers/net/ethernet/freescale/fec.h
+++ b/drivers/net/ethernet/freescale/fec.h
@@ -550,6 +550,7 @@ struct fec_enet_private {
 	struct	mii_bus *mii_bus;
 	int	mii_timeout;
 	int	mii_bus_share;
+	bool	miibus_up_failed;
 	uint	phy_speed;
 	phy_interface_t	phy_interface;
 	struct device_node *phy_node;
diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index 91321b6..d7a9d26 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -2967,12 +2967,14 @@ static inline bool fec_enet_irq_workaround(struct fec_enet_private *fep)
 
 	device_set_wakeup_enable(&ndev->dev, fep->wol_flag &
 				 FEC_WOL_FLAG_ENABLE);
+	fep->miibus_up_failed = false;
 
 	return 0;
 
 err_enet_mii_probe:
 	fec_enet_free_buffers(ndev);
 err_enet_alloc:
+	fep->miibus_up_failed = true;
 	if (!fep->mii_bus_share)
 		fec_enet_clk_enable(ndev, false);
 clk_enable:
@@ -3780,7 +3782,7 @@ static int __maybe_unused fec_suspend(struct device *dev)
 			enable_irq_wake(fep->wake_irq);
 		}
 		fec_enet_clk_enable(ndev, false);
-	} else if (fep->mii_bus_share && !ndev->phydev) {
+	} else if (fep->mii_bus_share && fep->miibus_up_failed && !ndev->phydev) {
 		fec_enet_clk_enable(ndev, false);
 		pinctrl_pm_select_sleep_state(&fep->pdev->dev);
 	}
@@ -3838,6 +3840,8 @@ static int __maybe_unused fec_resume(struct device *dev)
 		phy_start(ndev->phydev);
 	} else if (fep->mii_bus_share && !ndev->phydev) {
 		pinctrl_pm_select_default_state(&fep->pdev->dev);
+		fep->miibus_up_failed = true;
+		/* And then recovery mii bus */
 		fec_restore_mii_bus(ndev);
 	}
 	rtnl_unlock();
-- 
1.7.9.5

